### PR TITLE
Run JSpecify conformance tests with the Nullness Checker

### PR DESCRIPTION
**Pull request by rohan-shettyy**
- Extended JSpecify conformance test framework to run the conformance test suite on EISOP nullness checker
- Generates report to `checker/report/ConformanceTest-report.txt`
- Use standard distribution of Guava

### CODE REVIEW

File: checker/src/test/java/org/checkerframework/checker/test/junit/NullnessJSpecifyConformanceTest.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    private final ImmutableList<Path> deps;
+
+    private static final ImmutableList<String> TEST_OPTIONS =
+            ImmutableList.of(
+                    "-AassumePure",
+                    "-Adetailedmsgtext",
+                    "-AcheckImpl",
+                    "-AsuppressWarnings=conditional",
+                    "-Astrict",
+                    "-AshowTypes");
```
Comment (wmdietl) [inline checker/src/test/java/org/checkerframework/checker/test/junit/NullnessJSpecifyConformanceTest.java:40]:
```suggestion
                    "-AassumePure",
                    "-Adetailedmsgtext");
```

These options don't exist in EISOP: `"-AcheckImpl", "-Astrict", "-AshowTypes"`.
Not clear what the warning suppression would do.

File: checker/build.gradle [hunk 2]
```diff
@@ -1085,6 +1093,39 @@ task wpiPlumeLibTest(group: 'Verification') {
...
+}
+
+// Run JSpecify conformance tests
+task conformanceTests(type: Test, group: 'Verification') {
+    description 'Runs the conformance tests.'
+    include '**/*ConformanceTest.class'
+
+    shouldRunAfter test
+
+    environment "JSPECIFY_CONFORMANCE_TEST_MODE", "WRITE"
```
Comment (wmdietl) [inline checker/build.gradle:1112]: Why is this set to `WRITE`? That would overwrite the report file. Instead, we want to check for mismatches between the current execution and the existing report.

File: checker/build.gradle [hunk 3]
```diff
@@ -1085,6 +1093,39 @@ task wpiPlumeLibTest(group: 'Verification') {
     }
 }
 
+// Unzip JSpecify conformance tests using the artifact
+task unzipConformanceTestSuite(group: 'Verification') {
```
Comment (wmdietl) [inline checker/build.gradle:1097]: Please add `JSpecify` to these tasks and their descriptions.

File: checker/build.gradle [hunk 4]
```diff
@@ -1085,6 +1093,39 @@ task wpiPlumeLibTest(group: 'Verification') {
...
+    description 'Unzips the conformance test suite into the build directory.'
+    copy {
+        from zipTree(configurations.conformanceTestSuite.singleFile)
+        into "${buildDir}/conformance-test-suite"
```
Comment (wmdietl) [inline checker/build.gradle:1101]: How about `jspecify-conformance-tests`? That's closer to the artifact name.

File: checker/build.gradle (hunk 3 above)
Comment (wmdietl) [reply checker/build.gradle:1097]: Should this task have a `group`? Tasks with a group are shown in `gradlew tasks`. Is this task ever directly useful?

File: checker/build.gradle [hunk 5]
```diff
@@ -1085,6 +1093,39 @@ task wpiPlumeLibTest(group: 'Verification') {
...
+    }
+}
+
+// Run JSpecify conformance tests
+task conformanceTests(type: Test, group: 'Verification') {
+    description 'Runs the conformance tests.'
+    include '**/*ConformanceTest.class'
```
Comment (wmdietl) [inline checker/build.gradle:1108]: Make this the full name, to make it easier to find and consistent with the `exclude` above.

File: checker/build.gradle [hunk 6]
```diff
@@ -1085,6 +1093,39 @@ task wpiPlumeLibTest(group: 'Verification') {
...
+
+    // Conformance tests
+    inputs.files(unzipConformanceTestSuite)
+    inputs.files("${projectDir}/report/ConformanceTest-report.txt")
+
+    doFirst {
+        systemProperties([
+            "ConformanceTest.inputs": "${buildDir}/conformance-test-suite/assertions/org/jspecify/conformance/tests",
+            "ConformanceTest.report": "${projectDir}/report/ConformanceTest-report.txt",
+            "ConformanceTest.deps"  : fileTree("tests/nullness-jspecifyconformance/deps").join(":")
```
Comment (wmdietl) [inline checker/build.gradle:1122]: What dependencies is this picking up?

File: checker/build.gradle [hunk 7]
```diff
@@ -1085,6 +1093,39 @@ task wpiPlumeLibTest(group: 'Verification') {
...
+    environment "JSPECIFY_CONFORMANCE_TEST_MODE", "WRITE"
+
+    // Conformance tests
+    inputs.files(unzipConformanceTestSuite)
+    inputs.files("${projectDir}/report/ConformanceTest-report.txt")
+
+    doFirst {
+        systemProperties([
+            "ConformanceTest.inputs": "${buildDir}/conformance-test-suite/assertions/org/jspecify/conformance/tests",
+            "ConformanceTest.report": "${projectDir}/report/ConformanceTest-report.txt",
```
Comment (wmdietl) [inline checker/build.gradle:1121]: How about putting the file here: `"${projectDir}/src/test/resources/jspecify-conformance-test-report.txt"`?

File: checker/src/test/java/org/checkerframework/checker/test/junit/NullnessJSpecifyConformanceTest.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+/** An object to run the conformance tests against the EISOP Checker Framework. */
+public final class NullnessJSpecifyConformanceTest {
+
+    private final Path testDir;
```
Comment (wmdietl) [inline checker/src/test/java/org/checkerframework/checker/test/junit/NullnessJSpecifyConformanceTest.java:29]: The CI will complain about missing javadoc. Please add some description to all fields and methods.

File: checker/src/test/java/org/checkerframework/checker/test/junit/NullnessJSpecifyConformanceTest.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+                                        new FinalReportedFact(
+                                                Path.of(diagnostic.getFilename()),
+                                                diagnostic.getLineNumber(),
+                                                diagnostic.getMessage()))
+                        .collect(toImmutableSet());
+        return reportedFacts;
+    }
+}
+
+final class FinalReportedFact extends ReportedFact {
```
Comment (wmdietl) [inline checker/src/test/java/org/checkerframework/checker/test/junit/NullnessJSpecifyConformanceTest.java:96]: What is the `Final` referring to?

**Review by wmdietl — CHANGES_REQUESTED** [review]
Thanks for starting this!
CI is failing with:
````
A problem occurred configuring project ':checker'.
> The value for task ':checker:compileJava' property 'javaCompiler' is final and cannot be changed any further.
````

Is it working for you locally? I don't immediately see which of your changes causes this.

**Comment by rohan-shettyy** [comment]
Added a new commit addressing this feedback.

> Is it working for you locally? I don't immediately see which of your changes causes this.

Yes, this works locally. I am also not sure what causes the failure.

File: checker/build.gradle [hunk 10]
```diff
@@ -1085,6 +1093,38 @@ task wpiPlumeLibTest(group: 'Verification') {
...
+    copy {
+        from zipTree(configurations.jSpecifyConformanceTestSuite.singleFile)
+        into "${buildDir}/jspecify-conformance-tests"
+    }
+}
+
+// Run JSpecify conformance tests
+tasks.register('jSpecifyConformanceTests', Test) {
+    group = 'Verification'
+    description 'Runs the conformance tests.'
```
Comment (wmdietl) [inline checker/build.gradle:1108]:
```suggestion
    description 'Runs the JSpecify conformance tests.'
```

File: checker/build.gradle [hunk 11]
```diff
@@ -1085,6 +1093,38 @@ task wpiPlumeLibTest(group: 'Verification') {
     }
 }
 
+// Unzip JSpecify conformance tests using the artifact
+tasks.register('unzipJSpecifyConformanceTestSuite', Copy) {
```
Comment (wmdietl) [inline checker/build.gradle:1097]: I tried this different (maybe lazily evaluated) task syntax, but it didn't fix our CI problem. Feel free to undo those changes.

The offending line is [here](https://github.com/eisop/checker-framework/blob/e5bfb16b1076756288ce847f25d50df5d3a682c1/build.gradle#L393). From the documentation it sounds like that property should be settable multiple times...

Can you try minimizing what change introduces the problem? E.g. only have the tasks, but don't add the dependency on `test`, remove one of the tasks, etc. until you have a working CI.

**Review by wmdietl — COMMENTED** [review]
Thanks for the improvements!
I noticed that you made the PR from your forks `master` branch. You should keep your master clean and in sync with eisop/master. Create a local feature branch for each PR you'll create.

**Comment by rohan-shettyy** [comment]
Removing the tasks dependency and the `runJSpecifyConformanceTests` task individually does not remove the error. Attempting to remove the unzip task gives an error:

```
Execution failed for task ':checker:compileTestJava'.
> Could not resolve all files for configuration ':checker:testCompileClasspath'.
   > Could not resolve org.jspecify.conformance:conformance-test-framework:0.0.0-SNAPSHOT.
     Required by:
         project :checker
      > No matching variant of org.jspecify.conformance:conformance-test-framework:0.0.0-SNAPSHOT:20240207.191836-12 was found. The consumer was configured to find a library for use during compile-time, compatible with Java 8, preferably in the form of class files, preferably optimized for standard JVMs, and its dependencies declared externally but:
          - Variant 'apiElements' capability org.jspecify.conformance:conformance-test-framework:0.0.0-SNAPSHOT declares a library for use during compile-time, packaged as a jar, and its dependencies declared externally:
              - Incompatible because this component declares a component, compatible with Java 9 and the consumer needed a component, compatible with Java 8
              - Other compatible attribute:
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
          - Variant 'runtimeElements' capability org.jspecify.conformance:conformance-test-framework:0.0.0-SNAPSHOT declares a library for use during runtime, packaged as a jar, and its dependencies declared externally:
              - Incompatible because this component declares a component, compatible with Java 9 and the consumer needed a component, compatible with Java 8
              - Other compatible attribute:
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
```

Could this be an issue with the conformance-test-framework artifact?

**Comment by wmdietl** [comment]
> Could this be an issue with the conformance-test-framework artifact?

As just discussed, let's try adding a version check.
We don't have a 9+ variable [here](https://github.com/eisop/checker-framework/blob/e5bfb16b1076756288ce847f25d50df5d3a682c1/build.gradle#L41) yet, so add that there.

File: checker/src/test/resources/jspecify-conformance-test-report.txt [hunk 12]
```diff
@@ -0,0 +1,115 @@
+# 14 pass; 100 fail; 114 total; 12.3% score
```
Comment (wmdietl) [inline checker/src/test/resources/jspecify-conformance-test-report.txt:1]: This file is no longer needed, right?

File: checker/bin-devel/test-jspecify-conformance.sh [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+
+SCRIPTDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+# shellcheck disable=SC1090 # In newer shellcheck than 0.6.0, pass: "-P SCRIPTDIR" (literally)
+export ORG_GRADLE_PROJECT_useJdk17Compiler=true
+source "$SCRIPTDIR"/clone-related.sh
+
+PLUME_SCRIPTS="$SCRIPTDIR/.plume-scripts"
+"$PLUME_SCRIPTS/git-clone-related" eisop jspecify-conformance
+cd ../jspecify-conformance
+./gradlew clean test --console=plain -PcfLocal
```
Comment (wmdietl) [inline checker/bin-devel/test-jspecify-conformance.sh:17]: Line ending.

Comment (wmdietl) [inline checker/bin-devel/test-jspecify-conformance.sh:17]: `clean` isn't necessary.

File: build.gradle [hunk 14]
```diff
@@ -29,6 +29,7 @@ apply plugin: 'de.undercouch.download'
 
 // There is another `repositories { ... }` block below; if you change this one, change that one as well.
 repositories {
+    maven { url 'https://s01.oss.sonatype.org/content/repositories/snapshots'}
```
Comment (wmdietl) [inline build.gradle:32]: Changes in this file aren't necessary any more.

File: azure-pipelines.yml [hunk 15]
```diff
@@ -803,6 +804,17 @@ jobs:
     continueOnError: true
     displayName: test-plume-lib.sh
 
+- job: jspecify_conformance
+  dependsOn:
+   - canary_jobs
+  pool:
+    vmImage: 'ubuntu-latest'
+  steps:
```
Comment (wmdietl) [inline azure-pipelines.yml:812]:
```suggestion
    vmImage: 'ubuntu-latest'
  container: wmdietl/cf-ubuntu-jdk21:latest
  steps:
```

To be consistent with all the other builds.

File: azure-pipelines.yml [hunk 16]
```diff
@@ -803,6 +804,17 @@ jobs:
     continueOnError: true
     displayName: test-plume-lib.sh
 
+- job: jspecify_conformance
```
Comment (wmdietl) [inline azure-pipelines.yml:807]:
```suggestion
- job: jspecify_conformance_jdk21
```

Also change above. Okay to test with only one version, but follow the same naming convention to make it clear.

**Review by wmdietl — CHANGES_REQUESTED** [review]
Thanks!

**Pull request merged**